File: react-native/android/src/main/java/com/reactnativedictation/DictationModule.kt
package com.reactnativedictation

import android.Manifest
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import androidx.core.content.ContextCompat
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.facebook.react.modules.core.PermissionAwareActivity
import com.facebook.react.modules.core.PermissionListener

class DictationModule(reactContext: ReactApplicationContext) : 
    ReactContextBaseJavaModule(reactContext),
    PermissionListener {

    private var coordinator: DictationCoordinator? = null
    private var permissionPromise: Promise? = null
    private var pendingOptions: DictationOptions? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun getName(): String = "DictationModule"

    // Events that can be sent to JavaScript
    companion object {
        const val EVENT_ON_RESULT = "onResult"
        const val EVENT_ON_STATUS = "onStatus"
        const val EVENT_ON_AUDIO_LEVEL = "onAudioLevel"
        const val EVENT_ON_AUDIO_FILE = "onAudioFile"
        const val EVENT_ON_ERROR = "onError"

        const val PERMISSION_REQUEST_CODE = 1001
    }

    // MARK: - Bridge Methods

    @ReactMethod
    fun initialize(promise: Promise) {
        mainHandler.post {
            try {
                if (coordinator == null) {
                    coordinator = DictationCoordinator(
                        context = reactApplicationContext,
                        eventEmitter = this
                    )
                }
                coordinator?.initialize()
                promise.resolve(null)
            } catch (e: Exception) {
                promise.reject("INIT_ERROR", e.message, e)
            }
        }
    }

    @ReactMethod
    fun startListening(options: ReadableMap?, promise: Promise) {
        mainHandler.post {
            val coordinator = this.coordinator
            if (coordinator == null) {
                promise.reject("NOT_INITIALIZED", "Dictation service not initialized")
                return@post
            }

            // Parse options first
            val opts = parseOptions(options)

            // Check permission
            if (!hasRecordPermission()) {
                // Store the promise and options for use after permission grant
                permissionPromise = promise
                pendingOptions = opts
                requestRecordPermission()
                return@post
            }

            try {
                coordinator.startListening(opts)
                promise.resolve(null)
            } catch (e: Exception) {
                promise.reject("START_ERROR", e.message, e)
            }
        }
    }

    @ReactMethod
    fun stopListening(promise: Promise) {
        mainHandler.post {
            try {
                coordinator?.stopListening()
                promise.resolve(null)
            } catch (e: Exception) {
                promise.reject("STOP_ERROR", e.message, e)
            }
        }
    }

    @ReactMethod
    fun cancelListening(promise: Promise) {
        mainHandler.post {
            try {
                coordinator?.cancelListening()
                promise.resolve(null)
            } catch (e: Exception) {
                promise.reject("CANCEL_ERROR", e.message, e)
            }
        }
    }

    @ReactMethod
    fun getAudioLevel(promise: Promise) {
        try {
            val level = coordinator?.getAudioLevel() ?: 0.0f
            promise.resolve(level.toDouble())
        } catch (e: Exception) {
            promise.resolve(0.0)
        }
    }

    @ReactMethod
    fun normalizeAudio(sourcePath: String, promise: Promise) {
        Thread {
            try {
                val encoder = AudioEncoderManager(reactApplicationContext)
                val result = encoder.normalizeAudio(sourcePath)
                
                val response = Arguments.createMap().apply {
                    putString("canonicalPath", result.canonicalPath)
                    putDouble("durationMs", result.durationMs)
                    putDouble("sizeBytes", result.sizeBytes.toDouble())
                    putBoolean("wasReencoded", result.wasReencoded)
                }
                
                mainHandler.post { promise.resolve(response) }
            } catch (e: Exception) {
                mainHandler.post { promise.reject("NORMALIZE_ERROR", e.message, e) }
            }
        }.start()
    }

    // Required for NativeEventEmitter
    @ReactMethod
    fun addListener(eventName: String) {
        // Keep: Required for RN NativeEventEmitter
    }

    @ReactMethod
    fun removeListeners(count: Int) {
        // Keep: Required for RN NativeEventEmitter
    }

    // MARK: - Event Emission

    fun emitResult(text: String, isFinal: Boolean) {
        val params = Arguments.createMap().apply {
            putString("text", text)
            putBoolean("isFinal", isFinal)
        }
        sendEvent(EVENT_ON_RESULT, params)
    }

    fun emitStatus(status: String) {
        val params = Arguments.createMap().apply {
            putString("status", status)
        }
        sendEvent(EVENT_ON_STATUS, params)
    }

    fun emitAudioLevel(level: Float) {
        val params = Arguments.createMap().apply {
            putDouble("level", level.toDouble())
        }
        sendEvent(EVENT_ON_AUDIO_LEVEL, params)
    }

    fun emitAudioFile(
        path: String,
        durationMs: Double,
        fileSizeBytes: Long,
        sampleRate: Double,
        channelCount: Int,
        wasCancelled: Boolean
    ) {
        val params = Arguments.createMap().apply {
            putString("path", path)
            putDouble("durationMs", durationMs)
            putDouble("fileSizeBytes", fileSizeBytes.toDouble())
            putDouble("sampleRate", sampleRate)
            putInt("channelCount", channelCount)
            putBoolean("wasCancelled", wasCancelled)
        }
        sendEvent(EVENT_ON_AUDIO_FILE, params)
    }

    fun emitError(message: String, code: String? = null) {
        val params = Arguments.createMap().apply {
            putString("message", message)
            code?.let { putString("code", it) }
        }
        sendEvent(EVENT_ON_ERROR, params)
    }

    private fun sendEvent(eventName: String, params: WritableMap) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }

    // MARK: - Permissions

    private fun hasRecordPermission(): Boolean {
        return ContextCompat.checkSelfPermission(
            reactApplicationContext,
            Manifest.permission.RECORD_AUDIO
        ) == PackageManager.PERMISSION_GRANTED
    }

    private fun requestRecordPermission() {
        val activity = currentActivity as? PermissionAwareActivity
        if (activity == null) {
            // No activity available (e.g., app is backgrounded)
            // Reject the promise immediately with a descriptive error
            permissionPromise?.reject(
                "NO_ACTIVITY",
                "Cannot request permission: app activity is not available. Ensure the app is in the foreground."
            )
            permissionPromise = null
            pendingOptions = null
            return
        }
        
        activity.requestPermissions(
            arrayOf(Manifest.permission.RECORD_AUDIO),
            PERMISSION_REQUEST_CODE,
            this
        )
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ): Boolean {
        if (requestCode == PERMISSION_REQUEST_CODE) {
            val promise = permissionPromise
            val options = pendingOptions ?: DictationOptions()
            
            // Clear stored promise and options in all cases
            permissionPromise = null
            pendingOptions = null
            
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                // Permission granted, start listening with the stored options
                promise?.let { p ->
                    try {
                        coordinator?.startListening(options)
                        p.resolve(null)
                    } catch (e: Exception) {
                        p.reject("START_ERROR", e.message, e)
                    }
                }
            } else {
                // Permission denied
                promise?.reject(
                    "NOT_AUTHORIZED",
                    "Microphone permission denied"
                )
            }
            return true
        }
        return false
    }

    // MARK: - Options Parsing

    private fun parseOptions(map: ReadableMap?): DictationOptions {
        if (map == null) return DictationOptions()
        
        return DictationOptions(
            preserveAudio = map.getBooleanOrDefault("preserveAudio", false),
            preservedAudioFilePath = map.getStringOrNull("preservedAudioFilePath"),
            deleteAudioIfCancelled = map.getBooleanOrDefault("deleteAudioIfCancelled", true)
        )
    }

    private fun ReadableMap.getBooleanOrDefault(key: String, default: Boolean): Boolean {
        return if (hasKey(key)) getBoolean(key) else default
    }

    private fun ReadableMap.getStringOrNull(key: String): String? {
        return if (hasKey(key)) getString(key) else null
    }

    // MARK: - Cleanup

    override fun invalidate() {
        coordinator?.destroy()
        coordinator = null
        super.invalidate()
    }
}

data class DictationOptions(
    val preserveAudio: Boolean = false,
    val preservedAudioFilePath: String? = null,
    val deleteAudioIfCancelled: Boolean = true
)

File: react-native/ios/DictationModule.swift
import Foundation
import AVFoundation
import React

@objc(DictationModule)
class DictationModule: RCTEventEmitter {
    
    // MARK: - Properties
    
    private var coordinator: DictationCoordinator?
    private var hasListeners = false
    
    // MARK: - Module Setup
    
    override init() {
        super.init()
    }
    
    override func supportedEvents() -> [String]! {
        return [
            "onResult",
            "onStatus",
            "onAudioLevel",
            "onAudioFile",
            "onError"
        ]
    }
    
    override func startObserving() {
        hasListeners = true
    }
    
    override func stopObserving() {
        hasListeners = false
    }
    
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    // MARK: - Bridge Methods
    
    @objc func initialize(
        _ options: NSDictionary?,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        Task { @MainActor in
            do {
                if coordinator == nil {
                    coordinator = DictationCoordinator(eventEmitter: self)
                }
                
                // Parse locale from options (BCP-47 format, e.g., "en-US")
                let locale: Locale
                if let optionsDict = options as? [String: Any],
                   let localeString = optionsDict["locale"] as? String,
                   !localeString.isEmpty {
                    locale = Locale(identifier: localeString)
                } else {
                    locale = Locale.current
                }
                
                try await coordinator?.initialize(locale: locale)
                resolve(nil)
            } catch {
                let dictationError = DictationError.from(error)
                reject(dictationError.code, dictationError.localizedDescription, error)
            }
        }
    }
    
    @objc func startListening(
        _ options: NSDictionary?,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        Task { @MainActor in
            guard let coordinator = coordinator else {
                reject("NOT_INITIALIZED", "Dictation service not initialized. Call initialize() first.", nil)
                return
            }
            
            do {
                let parsedOptions = try DictationStartOptions.from(dictionary: options as? [String: Any])
                try await coordinator.startListening(options: parsedOptions)
                resolve(nil)
            } catch {
                let dictationError = DictationError.from(error)
                reject(dictationError.code, dictationError.localizedDescription, error)
            }
        }
    }
    
    @objc func stopListening(
        _ resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            do {
                try await coordinator?.stopListening()
                resolve(nil)
            } catch {
                let dictationError = DictationError.from(error)
                reject(dictationError.code, dictationError.localizedDescription, error)
            }
        }
    }
    
    @objc func cancelListening(
        _ resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            do {
                try await coordinator?.cancelListening()
                resolve(nil)
            } catch {
                let dictationError = DictationError.from(error)
                reject(dictationError.code, dictationError.localizedDescription, error)
            }
        }
    }
    
    @objc func getAudioLevel(
        _ resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        guard let coordinator = coordinator else {
            resolve(0.0)
            return
        }
        
        let level = coordinator.getAudioLevel()
        resolve(level)
    }
    
    @objc func normalizeAudio(
        _ sourcePath: String,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            do {
                let encoder = AudioEncoderManager()
                let result = try await encoder.normalizeAudio(sourcePath: sourcePath)
                
                let response: [String: Any] = [
                    "canonicalPath": result.canonicalPath,
                    "durationMs": Int(result.durationMs),
                    "sizeBytes": result.sizeBytes,
                    "wasReencoded": result.wasReencoded
                ]
                
                await MainActor.run {
                    resolve(response)
                }
            } catch {
                let normError = error as? NormalizationError ??
                    NormalizationError.encoderError("Unknown error: \(error.localizedDescription)")
                
                await MainActor.run {
                    reject(normError.code, normError.localizedDescription, error)
                }
            }
        }
    }
    
    // MARK: - Event Emission Helpers
    
    func emitResult(text: String, isFinal: Bool) {
        guard hasListeners else { return }
        
        sendEvent(withName: "onResult", body: [
            "text": text,
            "isFinal": isFinal
        ])
    }
    
    func emitStatus(_ status: String) {
        guard hasListeners else { return }
        
        sendEvent(withName: "onStatus", body: [
            "status": status
        ])
    }
    
    func emitAudioLevel(_ level: Float) {
        guard hasListeners else { return }
        
        sendEvent(withName: "onAudioLevel", body: [
            "level": level
        ])
    }
    
    func emitAudioFile(
        path: String,
        durationMs: Double,
        fileSizeBytes: Int64,
        sampleRate: Double,
        channelCount: Int,
        wasCancelled: Bool
    ) {
        guard hasListeners else { return }
        
        sendEvent(withName: "onAudioFile", body: [
            "path": path,
            "durationMs": durationMs,
            "fileSizeBytes": fileSizeBytes,
            "sampleRate": sampleRate,
            "channelCount": channelCount,
            "wasCancelled": wasCancelled
        ])
    }
    
    func emitError(message: String, code: String? = nil) {
        guard hasListeners else { return }
        
        var body: [String: Any] = ["message": message]
        if let code = code {
            body["code"] = code
        }
        
        sendEvent(withName: "onError", body: body)
    }
    
    // MARK: - Cleanup
    
    deinit {
        coordinator = nil
    }
}

// MARK: - Options Parsing

struct DictationStartOptions {
    let preserveAudio: Bool
    let preservedAudioFilePath: String?
    let deleteAudioIfCancelled: Bool
    
    static func from(dictionary: [String: Any]?) throws -> DictationStartOptions {
        guard let dict = dictionary else {
            return DictationStartOptions(
                preserveAudio: false,
                preservedAudioFilePath: nil,
                deleteAudioIfCancelled: true
            )
        }
        
        return DictationStartOptions(
            preserveAudio: dict["preserveAudio"] as? Bool ?? false,
            preservedAudioFilePath: (dict["preservedAudioFilePath"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
            deleteAudioIfCancelled: dict["deleteAudioIfCancelled"] as? Bool ?? true
        )
    }
}
